Feat: Fix the Message Sign issue for ledger devices

dawnseeker8 opened the pull request:
**Description**

After rebase from `main` branch to `ledger-intergtation-main` branch, the whole ledger message sign including `eth_sign`, `personal_sign`, `eth_signTypedDataV1` to V4 all break, this PR include fix from above operation.

following tasks have been done:
1. add two missing patch files in current branch but exists in `main` branch (`patches/@metamask+signature-controller++@metamask+message-manager+7.0.1.patch`, `patches/@metamask+signature-controller+4.0.1.patch`)
2. add `rpcEvent` actions and reducer to track the difference stage of whilelisted RPC event stage, currently mainly for `signing` relative RPC usage..
3. modify `LedgerMessageSignModal.tsx` to subscribe the `signingEvents` in redux store. to perform dismissModal. 
4. refactory the RPCMethodMiddleware to inject the track RPC stage dispatch event in middleware code to listen to all whitelisted RPC methods.
5. fix the code errors in MessageSIgn.tsx and PersonalSign.tsx fils for navigation.

**Caution** unit tests for testing reducer and RPCMethodMiddleware.ts will be done in another unit test branch.

stanleyyconsensys committed:
feat: add ledger sign message flow

- update message sign, personal sign and typed sign method
- implement ledger sign flow on above case

stanleyyconsensys committed:
Merge branch 'feat/ledger-intergtation-main' into feat/implement-sign-message-for-ledger

Akaryatrh committed:
fix: change isExternalHardwareAccount to use proper value

stanleyyconsensys committed:
fix: use org signature flow

- dismiss org sign model for ledger
- rollback deferSetAsSigned
- use org keyring sign flow

stanleyyconsensys committed:
fix: lint fix

dawnseeker8 committed:
Feat: rebase `ledger-integration-main`

Merge branch 'ledger-intergtation-main' into implement-sign-message-for-ledger

dawnseeker8 committed:
feat: Rebase ledger-intergation-main

add modified `yarn.lock` and `PersonalSign.tsx` files

dawnseeker8 committed:
feat: backup a new version of personal sign

this version still have bug, need to fix later,

dawnseeker8 committed:
feat: backup a new version of personal sign

this version still have bug, need to fix later,

dawnseeker8 committed:
Merge branch 'feat/ledger-intergtation-main' into feat/implement-sign-message-for-ledger

dawnseeker8 committed:
feat: refacotry the code and added missing patch

following has been done:
1. add two missing patch files from main.
2. refactory the RPCMethodMiddleware code to use common handle code.
3. fix the code errors in MessageSIgn.tsx and PersonalSign.tsx fils.

dawnseeker8 committed:
feat: Refactory the code and refine the jsDoc

dawnseeker8 committed:
feat: Rewrite the jsDoc

stanleyyconsensys commented on app/components/UI/LedgerModals/LedgerMessageSignModal.tsx:
```diff
@@ -25,45 +29,46 @@ export const createLedgerMessageSignModalNavDetails =
+    // This requires the user to confirm on the ledger device
+    let rawSignature;
 
     onConfirmationComplete(true, rawSignature);
```
shall we update onConfirmationComplete, as rawSignature will be undefined anyway

dawnseeker8 replied:
i see the original code implemented by Zoltan, the rawSignature is optioinal in onConfirmationComplete(), let me test deviices when removing `rawSignature` in that method, see any unexpected behaviour or not.

stanleyyconsensys replied:
onConfirmationComplete is passed by app/util/hardwareWallet/signatureUtils.ts, and since we no longer has rawSignature until keyring signed, hence, i think we dont need it anymore, you can double check

dawnseeker8 replied:
i have checked remove rawSignature and change to use `onComfirmationComplete(true)` and it still work as expected, so i have removed the rawSignature in above code.

stanleyyconsensys commented on app/core/RPCMethods/RPCMethodMiddleware.ts:
```diff
@@ -151,6 +160,33 @@ export const checkActiveAccountAndChainId = async ({
+
+    return rawSig;
+  } catch (e) {
+    console.error(e);
```
console.error may needed to remove?

dawnseeker8 replied:
will remove later.

stanleyyconsensys commented on app/core/RPCMethods/RPCMethodMiddleware.ts:
```diff
@@ -526,17 +562,14 @@ export const getRpcMethodMiddleware = ({
 
-          res.result = rawSig;
-        } else {
-          res.result = AppConstants.ETH_SIGN_ERROR;
```
it look like ur generic method doesnt handle this

why not just have a try catch on the req method, hence to leave everything not change , doesnt it more easy ? as the logic will update in future too

dawnseeker8 replied:
Sorry, my mistake to remove the `} else {` probably from copy and parse the code.  anyway, my generic method will  only process the signing bit, so after i add the `} else {` the `eth_sign` will still work, the code will change as following
```Typescript
     if (req.params[1].length === 66 || req.params[1].length === 67) {
          await checkActiveAccountAndChainId({
            hostname,
            address: req.params[0].from,
            checkSelectedAddress: isMMSDK || isWalletConnect,
          });
          const signMessageFunc = SignatureController.newUnsignedMessage({
            data: req.params[1],
            from: req.params[0],
            ...pageMeta,
            origin: hostname,
          });

          res.result = await handleCommonSignFlow(signMessageFunc);
        } else {
          res.result = AppConstants.ETH_SIGN_ERROR;
          throw ethErrors.rpc.invalidParams(AppConstants.ETH_SIGN_ERROR);
        }
      ```

stanleyyconsensys commented on app/components/UI/SignatureRequest/types.ts:
```diff
@@ -16,7 +16,12 @@ export interface PageMeta {
   data: string;
   from: string;
-  metamaskId: string;
+  metamaskId?: string;
```
what is the reason to change to optional?

dawnseeker8 replied:
The reason is metamaskId is mandatory defined in params in `SignatureController.newUnsign<>Message(messageParam)`, if you look at `RPCMethodMiddleware.ts` file for difference sign method, the params inside signMsgCallback may not include metamaskId, that is why i define it optional, however, the best way is getting the param types directly from @Metamask-SignatureController, but our library version didnt export the type, and i saw you already define a duplicate one in this file, i just directly use it to solve the tsc error.

stanleyyconsensys replied:
i see, the type for MessageParams original was not used for RPCMethodMiddleware, but for approval request
until we move the sign function as a callback in RPCMethodMiddleware and then you have to define the type to bypass the type error

got it, but if we dont use that way (move the sign function as a callback) on RPCMethodMiddleware , just rmb to roll it back, thx

stanleyyconsensys commented on app/core/RPCMethods/RPCMethodMiddleware.ts:
```diff
@@ -151,6 +160,33 @@ export const checkActiveAccountAndChainId = async ({
+ */
+const handleCommonSignFlow = async (
+  signMsgCallback: (
+    messageParams: MessageParams,
```
for the type , you can refer to sign controller

dawnseeker8 replied:
I check the @Metamask-SignatureController library,  the current version didnt export the types including `MessageParams` and `OriginalRequest` , the only way is upgrade or downgrade library, i saw you already use the `MessageParams` in `SignatureRequest/types.ts`, that is why i directly modify there, and add a `OriginalRequest` type there.

stanleyyconsensys replied:
they did export , as they import from metamask/message-manager

import type {
  MessageParams,
  MessageParamsMetamask,
  PersonalMessageParams,
  PersonalMessageParamsMetamask,
  TypedMessageParams,
  TypedMessageParamsMetamask,
  AbstractMessageManager,
  AbstractMessage,
  MessageManagerState,
  AbstractMessageParams,
  AbstractMessageParamsMetamask,
  OriginalRequest,
  TypedMessage,
  PersonalMessage,
  Message,
} from '@metamask/message-manager';

stanleyyconsensys commented on app/components/UI/LedgerModals/LedgerMessageSignModal.tsx:
```diff
@@ -25,45 +29,46 @@ export const createLedgerMessageSignModalNavDetails =
+    if (
+      signMessageStage === SignMessageStageTypes.COMPLETE ||
+      signMessageStage === SignMessageStageTypes.ERROR
+    ) {
```
when we on unmount, shall we dismiss modal too?

dawnseeker8 replied:
dismissModal wiill run first before unmounted,  so we dont need to handle dismissModal again here in unmounted callback

dawnseeker8 committed:
feat: backup the working code before refactory

1. fix tsc error in reducer
2. fix the eth_sign missing logic due to mistake
3. based on stanely feedback, change some code based on his comments

dawnseeker8 committed:
feat: refactory the RPCMethodMiddleware

1. Reimplement the generic code to middle method
2. dipatch actions in middleware method to reduce impact to actual RPC
3. create a whiteListedRPCs which can trigger actions.

this commit didnt include the reducer refactory
to make event action more generic.

dawnseeker8 committed:
feat: give a default value to bypass lint error

stanleyyconsensys commented on app/core/RPCMethods/RPCMethodMiddleware.ts:
```diff
@@ -885,7 +899,20 @@ export const getRpcMethodMiddleware = ({
+
+    try {
+      isWhiteListedMethod &&
+        store.dispatch(setSignMessageStage(SignMessageStageTypes.REQUEST_SEND));
```
do you wanna change the reducer name and action name to more general purpose as we discussed? or you prefer stay with explicitly name for sign message

dawnseeker8 replied:
last commit is to make a backup of workable version for further change, as i mention in the commit comment, further reducer change will be carried out later.

dawnseeker8 committed:
Merge branch 'feat/ledger-intergtation-main' into feat/implement-sign-message-for-ledger

dawnseeker8 committed:
feat: Refactory signMesage reducer to rpcEvents

1. move whitelist code to reducer to centrally controlled.
2. rename `signMessage` actions and reducer to `rpcEvents`
3. change LedgerMessageSignModal ui to subscribe to different state obj

dawnseeker8 committed:
feat: remove the redundant signMessage action

dawnseeker8 committed:
feat: lint:fix and remove type OriginalRequest

dawnseeker8 committed:
Merge branch 'feat/ledger-intergtation-main' into feat/implement-sign-message-for-ledger
